File: onionskin/diff.py
"""The core comparison: what ink is on the edited page that isn't on the original.

The whole app rests on one asymmetry. Ink that appears in the edited document
but not the original is *addable* — a printer can lay it down on the sheet you
already have. Ink that appears in the original but not the edited version is
not removable; toner does not come off paper. So the removed mask is never
printed, but it is the single most valuable signal we have: if anything
disappeared from where it used to be, the layout reflowed and the sheet in the
user's hand no longer matches the document.
"""

from __future__ import annotations

from collections import deque
from dataclasses import dataclass, field

import numpy as np
from PIL import Image, ImageFilter

from .geometry import MM_PER_INCH, PageSize, px_to_mm

#: Pixels at or below this grey level count as ink. Anti-aliased glyph edges
#: run light, so this sits well above pure black.
DEFAULT_INK_THRESHOLD = 200

#: How far a mark may move and still count as "the same mark". Absorbs
#: sub-pixel layout jitter between two renders of near-identical content.
DEFAULT_TOLERANCE_MM = 0.12

#: Additions closer together than this are reported as one region, so a word
#: comes back as one box rather than five letters.
DEFAULT_GROUP_MM = 2.0

#: Specks smaller than this are dropped as rendering noise.
DEFAULT_MIN_REGION_MM2 = 0.05


@dataclass
class Region:
    """A rectangle of changed ink, in page-space millimetres."""

    x0_mm: float
    y0_mm: float
    x1_mm: float
    y1_mm: float
    ink_mm2: float
    px_bbox: tuple[int, int, int, int]  # x0, y0, x1, y1 in pixels

    @property
    def width_mm(self) -> float:
        return self.x1_mm - self.x0_mm

    @property
    def height_mm(self) -> float:
        return self.y1_mm - self.y0_mm

    @property
    def area_mm2(self) -> float:
        return self.width_mm * self.height_mm

    def padded(self, pad_mm: float, page: PageSize) -> "Region":
        return Region(
            x0_mm=max(0.0, self.x0_mm - pad_mm),
            y0_mm=max(0.0, self.y0_mm - pad_mm),
            x1_mm=min(page.width_mm, self.x1_mm + pad_mm),
            y1_mm=min(page.height_mm, self.y1_mm + pad_mm),
            ink_mm2=self.ink_mm2,
            px_bbox=self.px_bbox,
        )

    def to_dict(self) -> dict:
        return {
            "x_mm": round(self.x0_mm, 2),
            "y_mm": round(self.y0_mm, 2),
            "width_mm": round(self.width_mm, 2),
            "height_mm": round(self.height_mm, 2),
            "ink_mm2": round(self.ink_mm2, 3),
        }


@dataclass
class PageDiff:
    """Everything learned by comparing one page of both documents."""

    index: int
    size: PageSize
    dpi: float
    added: np.ndarray = field(repr=False)
    removed: np.ndarray = field(repr=False)
    added_px: int = 0
    removed_px: int = 0
    added_regions: list[Region] = field(default_factory=list)
    removed_regions: list[Region] = field(default_factory=list)

    def release(self) -> None:
        """Drop the pixel masks, keeping every derived measurement.

        A page at 400 dpi is a 13-megapixel mask; holding two of them per page
        for a long document is the difference between a few hundred megabytes
        and a few kilobytes. Counts and regions are computed up front so
        nothing downstream needs the pixels back.
        """
        empty = np.zeros((0, 0), dtype=bool)
        self.added = empty
        self.removed = empty

    @property
    def px_area_mm2(self) -> float:
        side = MM_PER_INCH / self.dpi
        return side * side

    @property
    def added_ink_mm2(self) -> float:
        return self.added_px * self.px_area_mm2

    @property
    def removed_ink_mm2(self) -> float:
        return self.removed_px * self.px_area_mm2

    @property
    def has_additions(self) -> bool:
        return bool(self.added_regions)

    @property
    def bounds_mm(self) -> tuple[float, float, float, float] | None:
        """Bounding box of every addition on the page."""
        if not self.added_regions:
            return None
        return (
            min(r.x0_mm for r in self.added_regions),
            min(r.y0_mm for r in self.added_regions),
            max(r.x1_mm for r in self.added_regions),
            max(r.y1_mm for r in self.added_regions),
        )

    def to_dict(self) -> dict:
        return {
            "page": self.index + 1,
            "size_mm": [round(self.size.width_mm, 1), round(self.size.height_mm, 1)],
            "added_ink_mm2": round(self.added_ink_mm2, 2),
            "removed_ink_mm2": round(self.removed_ink_mm2, 2),
            "added_regions": [r.to_dict() for r in self.added_regions],
            "removed_region_count": len(self.removed_regions),
        }


def _ink_mask(gray: np.ndarray, threshold: int) -> np.ndarray:
    return gray <= threshold


def _dilate(mask: np.ndarray, radius_px: int) -> np.ndarray:
    """Grow a boolean mask by ``radius_px`` using Pillow's C max filter."""
    if radius_px <= 0:
        return mask
    size = 2 * radius_px + 1
    img = Image.fromarray((mask * 255).astype(np.uint8), mode="L")
    grown = img.filter(ImageFilter.MaxFilter(size))
    return np.asarray(grown, dtype=np.uint8) > 0


def label_regions(
    mask: np.ndarray,
    dpi: float,
    group_mm: float = DEFAULT_GROUP_MM,
    min_area_mm2: float = DEFAULT_MIN_REGION_MM2,
) -> list[Region]:
    """Group set pixels into regions with exact bounding boxes.

    Connectivity is resolved on a coarse grid of ``group_mm`` cells rather than
    per pixel: at 400 dpi an A4 page is 13 megapixels, but only ~62k cells, so
    the flood fill stays cheap while still merging the letters of a word into
    one box. Bounding boxes are then measured back at full resolution, cell by
    cell, so nothing is rounded up to the grid.
    """
    if not mask.any():
        return []

    h, w = mask.shape
    cell = max(1, int(round(group_mm * dpi / MM_PER_INCH)))
    gh = (h + cell - 1) // cell
    gw = (w + cell - 1) // cell

    padded = np.zeros((gh * cell, gw * cell), dtype=bool)
    padded[:h, :w] = mask
    grid = padded.reshape(gh, cell, gw, cell).any(axis=(1, 3))

    px_mm2 = (MM_PER_INCH / dpi) ** 2
    seen = np.zeros_like(grid)
    regions: list[Region] = []

    for gy in range(gh):
        for gx in range(gw):
            if not grid[gy, gx] or seen[gy, gx]:
                continue
            seen[gy, gx] = True
            queue = deque([(gy, gx)])
            cells: list[tuple[int, int]] = []
            while queue:
                cy, cx = queue.popleft()
                cells.append((cy, cx))
                for dy in (-1, 0, 1):
                    for dx in (-1, 0, 1):
                        ny, nx = cy + dy, cx + dx
                        if 0 <= ny < gh and 0 <= nx < gw:
                            if grid[ny, nx] and not seen[ny, nx]:
                                seen[ny, nx] = True
                                queue.append((ny, nx))

            x0 = y0 = 1 << 30
            x1 = y1 = -1
            ink_px = 0
            for cy, cx in cells:
                r0, c0 = cy * cell, cx * cell
                sub = mask[r0 : min(r0 + cell, h), c0 : min(c0 + cell, w)]
                if not sub.any():
                    continue
                rows = np.flatnonzero(sub.any(axis=1))
                cols = np.flatnonzero(sub.any(axis=0))
                y0 = min(y0, r0 + int(rows[0]))
                y1 = max(y1, r0 + int(rows[-1]) + 1)
                x0 = min(x0, c0 + int(cols[0]))
                x1 = max(x1, c0 + int(cols[-1]) + 1)
                ink_px += int(sub.sum())

            if x1 < 0:
                continue
            ink_mm2 = ink_px * px_mm2
            if ink_mm2 < min_area_mm2:
                continue
            regions.append(
                Region(
                    x0_mm=px_to_mm(x0, dpi),
                    y0_mm=px_to_mm(y0, dpi),
                    x1_mm=px_to_mm(x1, dpi),
                    y1_mm=px_to_mm(y1, dpi),
                    ink_mm2=ink_mm2,
                    px_bbox=(x0, y0, x1, y1),
                )
            )

    regions.sort(key=lambda r: (round(r.y0_mm, 1), r.x0_mm))
    return regions


def diff_page(
    old_gray: np.ndarray,
    new_gray: np.ndarray,
    size: PageSize,
    dpi: float,
    index: int = 0,
    ink_threshold: int = DEFAULT_INK_THRESHOLD,
    tolerance_mm: float = DEFAULT_TOLERANCE_MM,
    group_mm: float = DEFAULT_GROUP_MM,
    min_region_mm2: float = DEFAULT_MIN_REGION_MM2,
) -> PageDiff:
    """Compare one rendered page against another.

    Each mask is taken against the *dilated* opposite: a glyph that shifted by
    a fraction of a pixel between renders would otherwise leave a hairline
    outline in the delta, and we would print a ghost of text that is already on
    the sheet.
    """
    h = min(old_gray.shape[0], new_gray.shape[0])
    w = min(old_gray.shape[1], new_gray.shape[1])
    old_gray = old_gray[:h, :w]
    new_gray = new_gray[:h, :w]

    old_ink = _ink_mask(old_gray, ink_threshold)
    new_ink = _ink_mask(new_gray, ink_threshold)

    radius = int(round(tolerance_mm * dpi / MM_PER_INCH))
    added = new_ink & ~_dilate(old_ink, radius)
    removed = old_ink & ~_dilate(new_ink, radius)

    return PageDiff(
        index=index,
        size=size,
        dpi=dpi,
        added=added,
        removed=removed,
        added_px=int(added.sum()),
        removed_px=int(removed.sum()),
        added_regions=label_regions(added, dpi, group_mm, min_region_mm2),
        removed_regions=label_regions(removed, dpi, group_mm, min_region_mm2),
    )


def blank_diff(size: PageSize, dpi: float, index: int) -> PageDiff:
    """A diff for a page that exists in only one of the two documents."""
    w, h = size.px_size(dpi)
    empty = np.zeros((h, w), dtype=bool)
    return PageDiff(index=index, size=size, dpi=dpi, added=empty, removed=empty)

File: rust/src/diff.rs
//! The core comparison: what ink is on the edited page that is not on the
//! original.
//!
//! The whole app rests on one asymmetry. Ink that appears in the edited
//! document but not the original is *addable* — a printer can lay it down on
//! the sheet you already have. Ink that appears in the original but not the
//! edited version is not removable; toner does not come off paper. So the
//! removed mask is never printed, but it is the single most valuable signal
//! there is: if anything disappeared from where it used to be, the layout
//! reflowed and the sheet in someone's hand no longer matches the document.

use crate::geometry::{px_to_mm, PageSize, MM_PER_INCH};

/// Pixels at or below this grey level count as ink. Anti-aliased glyph edges
/// run light, so this sits well above pure black.
pub const DEFAULT_INK_THRESHOLD: u8 = 200;

/// How far a mark may move and still count as "the same mark". Absorbs
/// sub-pixel layout jitter between two renders of near-identical content.
pub const DEFAULT_TOLERANCE_MM: f64 = 0.12;

/// Additions closer together than this are reported as one region, so a word
/// comes back as one box rather than five letters.
pub const DEFAULT_GROUP_MM: f64 = 2.0;

/// Specks smaller than this are dropped as rendering noise.
pub const DEFAULT_MIN_REGION_MM2: f64 = 0.05;

/// A boolean image, one bit of meaning per pixel.
#[derive(Debug, Clone, PartialEq)]
pub struct Mask {
    pub width: usize,
    pub height: usize,
    pub bits: Vec<bool>,
}

impl Mask {
    pub fn blank(width: usize, height: usize) -> Mask {
        Mask {
            width,
            height,
            bits: vec![false; width * height],
        }
    }

    #[inline]
    pub fn get(&self, x: usize, y: usize) -> bool {
        self.bits[y * self.width + x]
    }

    #[inline]
    pub fn set(&mut self, x: usize, y: usize, on: bool) {
        self.bits[y * self.width + x] = on;
    }

    pub fn count(&self) -> usize {
        self.bits.iter().filter(|b| **b).count()
    }

    pub fn any(&self) -> bool {
        self.bits.iter().any(|b| *b)
    }

    /// Free the pixels, keeping the shape.
    ///
    /// A page at 400 dpi is a thirteen-megapixel mask, and holding two per page
    /// for a long document is the difference between a few hundred megabytes
    /// and a few kilobytes. Everything downstream needs is measured before this
    /// is called.
    pub fn release(&mut self) {
        self.bits = Vec::new();
    }
}

/// A rectangle of changed ink, in page-space millimetres.
#[derive(Debug, Clone, Copy, PartialEq, serde::Serialize)]
pub struct Region {
    pub x0_mm: f64,
    pub y0_mm: f64,
    pub x1_mm: f64,
    pub y1_mm: f64,
    pub ink_mm2: f64,
    /// The same box in pixels: x0, y0, x1, y1, with x1/y1 exclusive.
    #[serde(skip)]
    pub px_bbox: (usize, usize, usize, usize),
}

impl Region {
    pub fn width_mm(&self) -> f64 {
        self.x1_mm - self.x0_mm
    }
    pub fn height_mm(&self) -> f64 {
        self.y1_mm - self.y0_mm
    }
    pub fn area_mm2(&self) -> f64 {
        self.width_mm() * self.height_mm()
    }

    /// The same region grown by `pad_mm`, clipped to the page.
    pub fn padded(&self, pad_mm: f64, page: PageSize) -> Region {
        Region {
            x0_mm: (self.x0_mm - pad_mm).max(0.0),
            y0_mm: (self.y0_mm - pad_mm).max(0.0),
            x1_mm: (self.x1_mm + pad_mm).min(page.width_mm),
            y1_mm: (self.y1_mm + pad_mm).min(page.height_mm),
            ink_mm2: self.ink_mm2,
            px_bbox: self.px_bbox,
        }
    }
}

/// Everything learned by comparing one page of both documents.
#[derive(Debug, Clone)]
pub struct PageDiff {
    pub index: usize,
    pub size: PageSize,
    pub dpi: f64,
    pub added: Mask,
    pub removed: Mask,
    pub added_px: usize,
    pub removed_px: usize,
    pub added_regions: Vec<Region>,
    pub removed_regions: Vec<Region>,
}

impl PageDiff {
    /// A diff for a page that exists in only one of the two documents.
    pub fn blank(size: PageSize, dpi: f64, index: usize) -> PageDiff {
        let (w, h) = size.px_size(dpi);
        PageDiff {
            index,
            size,
            dpi,
            added: Mask::blank(w as usize, h as usize),
            removed: Mask::blank(w as usize, h as usize),
            added_px: 0,
            removed_px: 0,
            added_regions: Vec::new(),
            removed_regions: Vec::new(),
        }
    }

    /// Drop the pixel masks, keeping every derived measurement.
    pub fn release(&mut self) {
        self.added.release();
        self.removed.release();
    }

    pub fn px_area_mm2(&self) -> f64 {
        let side = MM_PER_INCH / self.dpi;
        side * side
    }

    pub fn added_ink_mm2(&self) -> f64 {
        self.added_px as f64 * self.px_area_mm2()
    }

    pub fn removed_ink_mm2(&self) -> f64 {
        self.removed_px as f64 * self.px_area_mm2()
    }

    pub fn has_additions(&self) -> bool {
        !self.added_regions.is_empty()
    }

    /// Bounding box of every addition on the page.
    pub fn bounds_mm(&self) -> Option<(f64, f64, f64, f64)> {
        let first = self.added_regions.first()?;
        let mut bounds = (first.x0_mm, first.y0_mm, first.x1_mm, first.y1_mm);
        for region in &self.added_regions[1..] {
            bounds.0 = bounds.0.min(region.x0_mm);
            bounds.1 = bounds.1.min(region.y0_mm);
            bounds.2 = bounds.2.max(region.x1_mm);
            bounds.3 = bounds.3.max(region.y1_mm);
        }
        Some(bounds)
    }
}

/// Which pixels of a greyscale page are ink.
pub fn ink_mask(gray: &[u8], width: usize, height: usize, threshold: u8) -> Mask {
    Mask {
        width,
        height,
        bits: gray[..width * height]
            .iter()
            .map(|v| *v <= threshold)
            .collect(),
    }
}

/// Grow a mask by `radius_px` in every direction.
///
/// Dilation by a square is *separable*: growing by r horizontally and then by r
/// vertically gives the same result as one (2r+1)² window, but costs O(r)
/// passes instead of O(r²) comparisons per pixel. On a thirteen-megapixel page
/// that is the difference between the better part of a second and a few
/// milliseconds — this is the hottest operation in the whole app, since it runs
/// twice per page.
pub fn dilate(mask: &Mask, radius_px: usize) -> Mask {
    if radius_px == 0 {
        return mask.clone();
    }
    let (w, h) = (mask.width, mask.height);

    let mut grown = mask.clone();
    for shift in 1..=radius_px {
        if shift >= w {
            break;
        }
        for y in 0..h {
            let row = y * w;
            for x in shift..w {
                // Left neighbour reaches right, and vice versa.
                if mask.bits[row + x - shift] {
                    grown.bits[row + x] = true;
                }
                if mask.bits[row + x] {
                    grown.bits[row + x - shift] = true;
                }
            }
        }
    }

    let mut out = grown.clone();
    for shift in 1..=radius_px {
        if shift >= h {
            break;
        }
        for y in shift..h {
            let (here, above) = (y * w, (y - shift) * w);
            for x in 0..w {
                if grown.bits[above + x] {
                    out.bits[here + x] = true;
                }
                if grown.bits[here + x] {
                    out.bits[above + x] = true;
                }
            }
        }
    }
    out
}

/// Group set pixels into regions with exact bounding boxes.
///
/// Connectivity is resolved on a coarse grid of `group_mm` cells rather than
/// per pixel: at 400 dpi an A4 page is thirteen megapixels but only about
/// sixty thousand cells, so the flood fill stays cheap while still merging the
/// letters of a word into one box. Bounding boxes are then measured back at
/// full resolution, cell by cell, so nothing is rounded up to the grid.
pub fn label_regions(mask: &Mask, dpi: f64, group_mm: f64, min_area_mm2: f64) -> Vec<Region> {
    if mask.bits.is_empty() || !mask.any() {
        return Vec::new();
    }
    let (w, h) = (mask.width, mask.height);
    let cell = ((group_mm * dpi / MM_PER_INCH).round() as usize).max(1);
    let gh = h.div_ceil(cell);
    let gw = w.div_ceil(cell);

    // Which coarse cells hold any ink at all.
    let mut grid = vec![false; gw * gh];
    for y in 0..h {
        let row = y * w;
        let gy = y / cell;
        for x in 0..w {
            if mask.bits[row + x] {
                grid[gy * gw + x / cell] = true;
            }
        }
    }

    let px_mm2 = (MM_PER_INCH / dpi).powi(2);
    let mut seen = vec![false; gw * gh];
    let mut regions: Vec<Region> = Vec::new();
    let mut queue: std::collections::VecDeque<(usize, usize)> = std::collections::VecDeque::new();
    let mut cells: Vec<(usize, usize)> = Vec::new();

    for gy in 0..gh {
        for gx in 0..gw {
            if !grid[gy * gw + gx] || seen[gy * gw + gx] {
                continue;
            }
            seen[gy * gw + gx] = true;
            queue.clear();
            cells.clear();
            queue.push_back((gy, gx));

            while let Some((cy, cx)) = queue.pop_front() {
                cells.push((cy, cx));
                // Eight-connected: a word's letters touch diagonally as often
                // as squarely once the grid is this coarse.
                for dy in -1i64..=1 {
                    for dx in -1i64..=1 {
                        let ny = cy as i64 + dy;
                        let nx = cx as i64 + dx;
                        if ny < 0 || nx < 0 || ny >= gh as i64 || nx >= gw as i64 {
                            continue;
                        }
                        let (ny, nx) = (ny as usize, nx as usize);
                        if grid[ny * gw + nx] && !seen[ny * gw + nx] {
                            seen[ny * gw + nx] = true;
                            queue.push_back((ny, nx));
                        }
                    }
                }
            }

            // Back to full resolution for the box itself.
            let (mut x0, mut y0) = (usize::MAX, usize::MAX);
            let (mut x1, mut y1) = (0usize, 0usize);
            let mut ink_px = 0usize;
            let mut found = false;

            for &(cy, cx) in &cells {
                let r0 = cy * cell;
                let c0 = cx * cell;
                for y in r0..(r0 + cell).min(h) {
                    let row = y * w;
                    for x in c0..(c0 + cell).min(w) {
                        if !mask.bits[row + x] {
                            continue;
                        }
                        found = true;
                        ink_px += 1;
                        x0 = x0.min(x);
                        y0 = y0.min(y);
                        x1 = x1.max(x + 1);
                        y1 = y1.max(y + 1);
                    }
                }
            }
            if !found {
                continue;
            }
            let ink_mm2 = ink_px as f64 * px_mm2;
            if ink_mm2 < min_area_mm2 {
                continue;
            }
            regions.push(Region {
                x0_mm: px_to_mm(x0 as f64, dpi),
                y0_mm: px_to_mm(y0 as f64, dpi),
                x1_mm: px_to_mm(x1 as f64, dpi),
                y1_mm: px_to_mm(y1 as f64, dpi),
                ink_mm2,
                px_bbox: (x0, y0, x1, y1),
            });
        }
    }

    // Reading order, with y rounded so that two boxes on the same line of text
    // sort left to right rather than by a hundredth of a millimetre.
    regions.sort_by(|a, b| {
        let ay = (a.y0_mm * 10.0).round();
        let by = (b.y0_mm * 10.0).round();
        ay.partial_cmp(&by)
            .unwrap_or(std::cmp::Ordering::Equal)
            .then(
                a.x0_mm
                    .partial_cmp(&b.x0_mm)
                    .unwrap_or(std::cmp::Ordering::Equal),
            )
    });
    regions
}

/// How a page is compared.
#[derive(Debug, Clone, Copy)]
pub struct DiffOptions {
    pub ink_threshold: u8,
    pub tolerance_mm: f64,
    pub group_mm: f64,
    pub min_region_mm2: f64,
}

impl Default for DiffOptions {
    fn default() -> Self {
        Self {
            ink_threshold: DEFAULT_INK_THRESHOLD,
            tolerance_mm: DEFAULT_TOLERANCE_MM,
            group_mm: DEFAULT_GROUP_MM,
            min_region_mm2: DEFAULT_MIN_REGION_MM2,
        }
    }
}

/// Compare one rendered page against another.
///
/// Each mask is taken against the *dilated* opposite: a glyph that shifted by a
/// fraction of a pixel between renders would otherwise leave a hairline outline
/// in the delta, and Onionskin would print a ghost of text already on the sheet.
#[allow(clippy::too_many_arguments)]
pub fn diff_page(
    old_gray: &[u8],
    old_size: (usize, usize),
    new_gray: &[u8],
    new_size: (usize, usize),
    size: PageSize,
    dpi: f64,
    index: usize,
    options: &DiffOptions,
) -> PageDiff {
    // Two renders of the same page should agree on their pixel size, but a
    // rounding difference of one pixel is not worth refusing over.
    let w = old_size.0.min(new_size.0);
    let h = old_size.1.min(new_size.1);

    let crop = |gray: &[u8], from: (usize, usize)| -> Vec<u8> {
        if from.0 == w && from.1 == h {
            return gray[..w * h].to_vec();
        }
        let mut out = Vec::with_capacity(w * h);
        for y in 0..h {
            out.extend_from_slice(&gray[y * from.0..y * from.0 + w]);
        }
        out
    };
    let old_gray = crop(old_gray, old_size);
    let new_gray = crop(new_gray, new_size);

    let old_ink = ink_mask(&old_gray, w, h, options.ink_threshold);
    let new_ink = ink_mask(&new_gray, w, h, options.ink_threshold);

    let radius = (options.tolerance_mm * dpi / MM_PER_INCH).round() as usize;
    let old_grown = dilate(&old_ink, radius);
    let new_grown = dilate(&new_ink, radius);

    let mut added = Mask::blank(w, h);
    let mut removed = Mask::blank(w, h);
    for i in 0..w * h {
        added.bits[i] = new_ink.bits[i] && !old_grown.bits[i];
        removed.bits[i] = old_ink.bits[i] && !new_grown.bits[i];
    }

    PageDiff {
        index,
        size,
        dpi,
        added_px: added.count(),
        removed_px: removed.count(),
        added_regions: label_regions(&added, dpi, options.group_mm, options.min_region_mm2),
        removed_regions: label_regions(&removed, dpi, options.group_mm, options.min_region_mm2),
        added,
        removed,
    }
}

#[cfg(test)]
mod tests;
